### PR TITLE
Implement #4397, Internal Bomb Bay quirk, plus various fixes and improvements

### LINKED ISSUE #4397: RFE: Implement Internal Bomb Bay quirk
The Internal Bomb Bay quirk, described on page 227 of *Campaign Operations*, allows units, including fighters, to load additional bombs inside their cargo bays. However, dropping those bombs incurs a risk of explosion from ground fire on a 2d6 roll of 10+.

---

As a stopgap, the most simple implementation of this could simply increase the bomb capacity of each fighter by the number of tons of cargo they have (rounded down).

A more comprehensive implementation would require specifying which bombs are loaded where, and which ones are being dropped during combat, to determine whether the explosion risk is present that turn.

### PR DESCRIPTION
**Pull request by Sleet01**
This has some pretty big changes:

### New gameplay features:

1. Aerospace fighters, Conventional Fighters, Fixed Wing Support aircraft, Small Craft, and Dropships with the "Internal Bomb Bay" quirk can now mount bomb-type munitions internally.  The Bombs section of the configuration dialog will indicate Internal and/or External bomb options as appropriate.  When bombing, the user will be prompted separately for Internal and External bomb stores to utilize (this is necessary for accounting reasons, and to keep the UI simple).
2. Units that utilize their internal stores to make attacks (or to TAG, if a TAG pod is mounted) chance ground fire touching off any other stores in their internal bays: on a 10+ roll at the end of a round in which a unit has taken damage from ground fire, all remaining internal bombs explode, dealing their full damage directly to SI!

### Changes/Additions

1. Re-organized Aero class hierarchy slightly:
```
|Entity
|- Aero
    |- TeleMissile
    |- AeroSpaceFighter                 // Inserted between Aero and subclasses
    |   |- ConventionalFighter
    |   |   |- FixedWingSupport
    |   |- FighterSquadron             // Now made up of ASFs, not base Aeros
    |- SmallCraft
    |   |- Dropship
    |   |- EscapePods                     // This should probably go under Aero, but out of scope for this PR.
    |- Jumpship
         |- Warship
         |- SpaceStation   
```
in service of reducing the number of classes that needed updates for IBB.  This should also allow reducing the code in Aero considerably and moving ASF-only functionality into that class, if desired.

2. Updated the Internal Bomb Bay quirk tooltip and text.
3. Added handling for Internal and External Bombs, including updates to configuration and bomb targeting dialogs.
4. Added IBB-related End Phase report, with relevant messages and explosion check / damage application.
5. Added ephemeral Cargo Bay generation for Aeros that mount the Cargo Bay miscellaneous equipment, as opposed to actual bays.  These are not saved, but are used when calculating bomb space for IBB (and, perhaps later, for cargo?).
6. Fixed "Invalid Design" checks for A) extra weight on ASFs with internal bombs and ephemeral bays, and B) "BA ECM" on Small Craft (MekHQ Issue [#1970](https://github.com/MegaMek/mekhq/issues/1970#issuecomment-1861759314)
7. Added accounting for ground-fire-only damage per entity; this is required for 4) above.
8. Added separate listing of Internal and External munitions for summary view and MUL saving/loading.
9. Added ability for VTOLs to bomb their current hex if a "bomb" hex was not added during the Move phase - this appears to have caused some confusion, and VTOLs should be able to bomb their final location in the Attack phase.
10. Fixed a timeout error that would cause Princess to hang when attempting to move ground units in the presence of a large number of bombs.  This may be the root of some of the recent Princess hangs we have not been able to track down; I will investigate further.

### TO-DO / Not Yet Implemented

1. Separately accounting for each bomb in each disparate cargo bay: Cargo Bays need to be upgraded to a real Class with internal logic and accounting for this to be feasible.  Currently we offer no way to account for location of bombs until they are instantiated after leaving the lobby; everything just gets thrown into "NOS".  This unfortunately means that cheesing bombers with multiple 6-bomb bays won't protect IBB users.
2. Counting cargo space filled with bombs as "used": see 1).
3. Effect of a Cargo crit on aerospace units mounting internal bomb stores and using IBB: forum question [here](https://bg.battletech.com/forums/index.php/topic,83510.msg1981761.html#msg1981761) is currently being researched; currently no effect.
4. Verification of Fighter Squadron bombing capabilities; this came up as a question during development and is apparently not yet settled.

### Testing

1. Ran MM tests in Idea
2. Tested configuration, saving, loading, and MUL reinforcement with all valid Internal Bomb Bay unit types.
3. Verified Princess operations with Internal, External, and mixed munition loads.

### Note:

I updated the JVM options in build.gradle to use 4GB of RAM and disable AWT grabbing (source of a lot of freezing while debugging AWT windows).  I can remove this if folks don't feel it's necessary for them.

Close #4397

### CODE REVIEW

File: megamek/src/megamek/MegaMek.java [hunk 1]
```diff
@@ -62,6 +62,7 @@
         Thread.setDefaultUncaughtExceptionHandler((thread, t) -> {
             LogManager.getLogger().error("Uncaught Exception Detected", t);
             final String name = t.getClass().getName();
+            final StackTraceElement[] stack = t.getStackTrace();
```
[the remark this answers is not in the document]
Comment (Sleet01) [reply megamek/src/megamek/MegaMek.java:65]: Shoot, this is for additional logging printout from the top-level `try/catch`, not necessary here.

**Comment by SJuliez** [comment]
Please note that this will require PRs for MML and MHQ.

**Comment by Sleet01** [comment]
> Have you tested MML/MHQ with these changes?

Only MML so far, and only for generating new bombers.  There is no provision for loading bombs in MML.

I will confirm MHQ compatibility as I can.

File: megamek/src/megamek/server/GameManager.java [hunk 2]
```diff
@@ -20871,6 +20872,88 @@
...
+            if (e.getGroundAttackedByThisTurn().isEmpty()) {
+                return vReport;
+            }
+
+            Aero b = (Aero) e;
+            Report r;
+
+            if (b.getUsedInternalBombs() > 0) {
+                int id = e.getId();
+                Vector<TargetRoll> rolls = new Vector<>();
```
[the remark this answers is not in the document]
Comment (Sleet01) [reply megamek/src/megamek/server/GameManager.java:20914]: Removed.

File: megamek/src/megamek/common/MULParser.java [hunk 3]
```diff
@@ -2217,17 +2217,24 @@
...
                     if (!type.isBlank() && !load.isBlank()) {
                         int bombType = BombType.getBombTypeFromInternalName(type);
                         if ((bombType <= BombType.B_NONE) || (bombType >= BombType.B_NUM)) {
                             continue;
                         }
 
-                        bombChoices[bombType] += Integer.parseInt(load);
-                        ((IBomber) entity).setBombChoices(bombChoices);
+                        if (internal) {
+                            intBombChoices[bombType] += Integer.parseInt(load);
```
[the remark this answers is not in the document]
Comment (Sleet01) [reply megamek/src/megamek/common/MULParser.java:2232]: Wrapped.

File: megamek/src/megamek/common/MULParser.java [hunk 4]
```diff
@@ -2217,17 +2217,24 @@
...
+                            ((IBomber) entity).setIntBombChoices(intBombChoices);
+                        } else {
+                            extBombChoices[bombType] += Integer.parseInt(load);
```
[the remark this answers is not in the document]
Comment (Sleet01) [reply megamek/src/megamek/common/MULParser.java:2235]: Wrapped.

File: megamek/src/megamek/common/AeroSpaceFighter.java [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+        }
+    }
+
+    /**
+     * Damage a capital fighter's weapons. WeaponGroups are damaged by critical hits.
+     * This matches up the individual fighter's weapons and critical slots and damages those
+     * for MHQ resolution
+     * @param loc - Int corresponding to the location struck
+     */
+    public void damageCapFighterWeapons(int loc) {
```
[the remark this answers is not in the document]
Comment (Sleet01) [reply megamek/src/megamek/common/AeroSpaceFighter.java:121]: Removed from Aero and modified only caller to use FighterSquadron class instead.

File: megamek/src/megamek/common/Aero.java [hunk 6]
```diff
@@ -477,33 +481,68 @@
...
+
+    public void setUsedInternalBombs(int b){
+        usedInternalBombs = b;
+    }
+
+    public void increaseUsedInternalBombs(int b){
+        usedInternalBombs += b;
+    }
+
+    public int getUsedInternalBombs() {
```
[the remark this answers is not in the document]
Comment (Sleet01) [reply megamek/src/megamek/common/Aero.java:544]: Annotated.

File: megamek/src/megamek/common/Aero.java [hunk 7]
```diff
@@ -477,33 +481,68 @@
...
-        return Math.max(0, t - (int) Math.ceil(getBombPoints() / 5.0));
+        // The base Aero cannot carry bombs so no MP reduction
+        return t;
+    }
+
+    public void setUsedInternalBombs(int b){
+        usedInternalBombs = b;
+    }
+
+    public void increaseUsedInternalBombs(int b){
```
[the remark this answers is not in the document]
Comment (Sleet01) [reply megamek/src/megamek/common/Aero.java:540]: Annotated.

File: megamek/src/megamek/common/Aero.java [hunk 8]
```diff
@@ -477,33 +481,68 @@
...
     }
 
     @Override
     public int reduceMPByBombLoad(int t) {
-        return Math.max(0, t - (int) Math.ceil(getBombPoints() / 5.0));
+        // The base Aero cannot carry bombs so no MP reduction
+        return t;
+    }
+
+    public void setUsedInternalBombs(int b){
```
[the remark this answers is not in the document]
Comment (Sleet01) [reply megamek/src/megamek/common/Aero.java:536]: Annotated.

File: megamek/src/megamek/common/Aero.java [hunk 9]
```diff
@@ -477,33 +481,68 @@
...
     public int getMaxBombPoints() {
-        return maxBombPoints;
+        return maxExtBombPoints + maxIntBombPoints;
+    }
+
+    public int getMaxIntBombPoints() {
+        return (hasQuirk(OptionsConstants.QUIRK_POS_INTERNAL_BOMB)) ? maxIntBombPoints : 0;
+    }
+
+    public int getMaxExtBombPoints() {
```
[the remark this answers is not in the document]
Comment (Sleet01) [reply megamek/src/megamek/common/Aero.java:491]: Annotated.

File: megamek/src/megamek/common/LandAirMech.java [hunk 10]
```diff
@@ -1083,31 +1087,65 @@
...
 
+    public void setUsedInternalBombs(int b){
+        // Do nothing; LAMs don't take internal bomb bay hits like this
+    }
+
+    public void increaseUsedInternalBombs(int b){
+        // Do nothing
+    }
+
+    public int getUsedInternalBombs() {
```
[the remark this answers is not in the document]
Comment (Sleet01) [reply megamek/src/megamek/common/LandAirMech.java:1132]: Annotated.

File: megamek/src/megamek/common/LandAirMech.java [hunk 11]
```diff
@@ -1083,31 +1087,65 @@
...
+        if (bc.length == intBombChoices.length) {
+            intBombChoices = bc.clone();
         }
     }
 
+    public void setUsedInternalBombs(int b){
+        // Do nothing; LAMs don't take internal bomb bay hits like this
+    }
+
+    public void increaseUsedInternalBombs(int b){
```
[the remark this answers is not in the document]
Comment (Sleet01) [reply megamek/src/megamek/common/LandAirMech.java:1128]: Annotated.

File: megamek/src/megamek/common/LandAirMech.java [hunk 12]
```diff
@@ -1083,31 +1087,65 @@
...
-    public void setBombChoices(int[] bc) {
-        if (bc.length == bombChoices.length) {
-            bombChoices = bc;
+    public void setIntBombChoices(int[] bc) {
+        if (bc.length == intBombChoices.length) {
+            intBombChoices = bc.clone();
         }
     }
 
+    public void setUsedInternalBombs(int b){
```
[the remark this answers is not in the document]
Comment (Sleet01) [reply megamek/src/megamek/common/LandAirMech.java:1124]: Annotated.

File: megamek/src/megamek/common/VTOL.java [hunk 13]
```diff
@@ -280,16 +301,29 @@
...
 
+    public void setUsedInternalBombs(int b){
+        // Do nothing
+    }
+
+    public void increaseUsedInternalBombs(int b){
+        // Do nothing
+    }
+
+    public int getUsedInternalBombs() {
```
[the remark this answers is not in the document]
Comment (Sleet01) [reply megamek/src/megamek/common/VTOL.java:312]: Annotated.

File: megamek/src/megamek/common/VTOL.java [hunk 14]
```diff
@@ -280,16 +301,29 @@
         return Math.max(0, (t - (int) this.getBombs().stream().filter(m -> (m.getUsableShotsLeft() > 0)).count()));
     }
 
+    public void setUsedInternalBombs(int b){
```
[the remark this answers is not in the document]
Comment (Sleet01) [reply megamek/src/megamek/common/VTOL.java:304]: Annotated.

File: megamek/src/megamek/server/GameManager.java [hunk 15]
```diff
@@ -20871,6 +20872,88 @@
...
 
+    /**
+     * Check all aircraft that may have used internal bomb bays for incidental explosions
+     * caused by ground fire.
+     * @return
+     */
+    private Vector<Report> resolveInternalBombHits() {
+        Vector<Report> vFullReport = new Vector<>();
+        vFullReport.add(new Report(5600, Report.PUBLIC));
+        for (Iterator<Entity> i = game.getEntities(); i.hasNext(); ) {
```
[the remark this answers is not in the document]
Comment (Sleet01) [reply megamek/src/megamek/server/GameManager.java:20883]: Replaced with `game.getEntitiesVector()`.

**Comment by kuronekochomusuke** [comment]
can you update the Advanced Search Unit Type tab layout to reflect the new class structure in TWAdvancedSearchPanel?  

indent the Conventional Fighter under Aerospace Fight, since it now a child,.

the layout was meant to follow the class hierarchy 

image

**Comment by kuronekochomusuke** [comment]
in MechSearchFilter this should be removed since there is a class now.

        if (mech.isAerospaceFighter()) {
            entityType = entityType | Entity.ETYPE_AEROSPACEFIGHTER;
        }

**Comment by NickAragua** [comment]
I'm not sure that your last merge went well, you went from 60 to 125 files

**Comment by Sleet01** [comment]
> I'm not sure that your last merge went well, you went from 60 to 125 files

No, it definitely did not -_-;

**Comment by SJuliez** [comment]
I'm seeing strange files in this, like the userdirhelp I wrote looking as if it was new, a new unit, gradle.build changes (intentional?). I'm not sure what to do about it. Can you look into this?

**Comment by Sleet01** [comment]
> I'm seeing strange files in this, like the userdirhelp I wrote looking as if it was new, a new unit, gradle.build changes (intentional?). I'm not sure what to do about it. Can you look into this?

The gradle change I mentioned in the commit comment: it enables more stable debugging of AWT objects but can be pulled out if folks don't like it.

The other stuff is likely due to rebasing and merging a change I'd made here on GitHub in the web UI back to the PR branch.  If work takes a long time I like to rebase to the latest master state often so there's less work to do when my PR gets pulled (if it gets pulled).
The "new" files, including that new ComStar unit, were all added to MegaMek after I created this PR; rebasing adds them to this branch's history but when the PR is pulled they will be discarded as unnecessary changes.  But rebasing ensures that my updates won't break due to more recent updates to the base code repo.

**Comment by Sleet01** [comment]
> I'm not sure that your last merge went well, you went from 60 to 125 files

Well, looking back, I think you're right - something is screwed up.
I'm going to try creating a new branch from the initial PR commit, rebase to latest master, and then apply the last few fixes.
If that works I'll close this PR and open a new one from that branch.
The MHQ PR can stay as-is; nothing there has been changed by this turmoil.

**Comment by Sleet01** [comment]
@NickAragua @Windchild292 @kuronekochomusuke @SJuliez @HammerGS 

Apologies for the churn.  I'm closing this PR in favor of a cleaner one, without all the extra files and broken merges:
https://github.com/MegaMek/megamek/pull/5003

**Pull request closed without merging**